opm/simulators/wells/WellInterfaceFluidSystem.cpp

```diff
@@ -80,6 +80,24 @@ calculateReservoirRates(SingleWellState& ws) const
     std::vector<double> voidage_rates(np, 0.0);
     rateConverter_.calcReservoirVoidageRates(fipreg, pvtRegionIdx_, surface_rates, voidage_rates);
     ws.reservoir_rates = voidage_rates;
+
+    // Compute total connection reservoir rate CVPR/CVIR
+    auto& perf_data = ws.perf_data;
+    const int num_perf_well = perf_data.size();
+    const auto& surf_perf_rates = perf_data.phase_rates;
+    auto& perf_rates = perf_data.rates;
+    for( int i = 0; i < num_perf_well; ++i ) {
+        std::vector<double> surface_rates_perf(np, 0.0);
+        perf_rates[i] = 0.0;
+        for (int p = 0; p < np; ++p) {
+            surface_rates_perf[p] = surf_perf_rates[np*i+p];
+        }
+        std::vector<double> voidage_rates_perf(np, 0.0);
+        rateConverter_.calcReservoirVoidageRates(fipreg, pvtRegionIdx_, surface_rates_perf, voidage_rates_perf);
+        for (int p = 0; p < np; ++p) {
+            perf_rates[i] += voidage_rates_perf[p];
+        }
+    }
```

Comment:
I've looked at this and I'm still not able to quite get to the bottom of what's happening with the test failures.  The differences between sequential and parallel runs are smaller than before, but still large enough to be a little concerning.  While looking into this, I rewrote the function to the one below which has the benefit of not unnecessarily copying rate data, and of not needing to update `perf_rates[i]` in a loop.  I would suggest you use this version too.

I will continue to look into this.

---

```C++
template <typename FluidSystem>
void
WellInterfaceFluidSystem<FluidSystem>::
calculateReservoirRates(SingleWellState& ws) const
{
    const int fipreg = 0; // not considering the region for now
    const int np = number_of_phases_;

    this->rateConverter_
        .calcReservoirVoidageRates(fipreg,
                                   this->pvtRegionIdx_,
                                   ws.surface_rates,
                                   ws.reservoir_rates);

    // Compute total connection reservoir rate CVPR/CVIR
    auto& perf_data = ws.perf_data;
    const auto num_perf_well = perf_data.size();
    const auto& surf_perf_rates = perf_data.phase_rates;
    for (auto i = 0*num_perf_well; i < num_perf_well; ++i) {
        const auto surface_rates_perf = std::vector<double>
            { surf_perf_rates.begin() + (i + 0)*np ,
              surf_perf_rates.begin() + (i + 1)*np };

        std::vector<double> voidage_rates_perf(np, 0.0);
        this->rateConverter_
            .calcReservoirVoidageRates(fipreg,
                                       this->pvtRegionIdx_,
                                       surface_rates_perf,
                                       voidage_rates_perf);

        perf_data.rates[i] =
            std::accumulate(voidage_rates_perf.begin(),
                            voidage_rates_perf.end(), 0.0);
    }
}
```

Reply:
Thanks, I will update the code.